src/ui/toast.js

```diff
@@ -53,121 +40,99 @@ export const toastImportantStyles = {
 
     /** @private @const {!HTMLDocument} */
-    this.doc_ = win.document;
+    this.doc_ = this.win_.document;
+
+    /** @private @const {!web-activities/activity-ports.ActivityPorts} */
+    this.activityPorts_ = deps.activities();
 
-    /** @private @const {!ToastSpecDef} */
-    this.spec_ = spec;
+    /** @private @const {!../runtime/deps.DepsDef} */
+    this.deps_ = deps;
 
-    /** @private @const {!FriendlyIframe} */
-    this.iframe_ = new FriendlyIframe(this.doc_, {'class': 'swg-toast'});
+    /** @private @const {{'source': string}} */
+    this.args_ = args;
 
-    /** @private {?Element} */
-    this.container_ = null;
+    /** @private @const {!HTMLIFrameElement} */
+    this.iframe_ =
+        /** @type {!HTMLIFrameElement} */ (
+            createElement(this.doc_, 'iframe', iframeAttributes));
 
-    setImportantStyles(this.iframe_.getElement(), toastImportantStyles);
-    setStyles(this.iframe_.getElement(), topFriendlyIframePositionStyles);
+    setImportantStyles(this.iframe_, toastImportantStyles);
+    setStyles(this.iframe_, topFriendlyIframePositionStyles);
+
+            /** @private @const {!Promise} */
+    this.ready_ = new Promise(resolve => {
+      this.iframe_.onload = resolve;
+    });
   }
 
-  /**
-   * Gets the attached iframe instance.
-   * @return {!FriendlyIframe}
-   */
-  getIframe() {
-    return this.iframe_;
+  getElement() {
+    return this.iframe;
   }
 
   /**
-   * Gets the Iframe element.
-   * @return {!HTMLIFrameElement}
+   * When promise is resolved.
+   * @return {!Promise}
    */
-  getElement() {
-    return this.iframe_.getElement();
+  whenReady() {
+    return this.ready_;
   }
 
   /**
    * Opens the notification toast.
    * @return {!Promise}
    */
   open() {
-    const iframe = this.iframe_;
-    if (iframe.isConnected()) {
-      throw new Error('Already opened');
-    }
-    this.doc_.body.appendChild(iframe.getElement());  // Fires onload.
-
-    return iframe.whenReady().then(() => this.buildIframe_());
-  }
+    this.doc_.body.appendChild(this.iframe_);  // Fires onload.
 
-  /**
-   * Closes the toast.
-   */
-  close() {
-    this.doc_.body.removeChild(this.iframe_.getElement());
+    return this.whenReady().then(() => {
+      this.buildToast_();
+      return this;
+    });
   }
 
-  /**
-   * Builds the iframe with content and the styling after iframe is loaded.
-   * @private
-   * @return {!Toast}
-   */
-  buildIframe_() {
-    const iframe = this.iframe_;
-    const iframeDoc = iframe.getDocument();
-    const iframeBody = iframe.getBody();
-
-    // Inject Google fonts in <HEAD> section of the iframe.
-    injectFontsLink(iframeDoc, googleFontsUrl);
-    injectStyleSheet(iframeDoc, TOAST_CSS);
-
-    this.addItems_(iframeDoc, iframeBody);
-
-    return this;
+  buildToast_() {
+    return this.activityPorts_.openIframe(
+        this.iframe_, feUrl('/toastiframe'),
+        feArgs({
+          'publicationId': this.deps_.pageConfig().getPublicationId(),
```

Comment:
I think we should just pass these all from outside as constructor args.

Reply:
Done